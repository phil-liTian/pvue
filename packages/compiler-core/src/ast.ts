export type TemplateChildNode = any

export enum NodeTypes {
  ROOT,
  ELEMENT,
  TEXT,
}

export const locStub: SourceLocation = {
  start: { line: 1, column: 1, offset: 0 },
  end: { line: 1, column: 1, offset: 0 },
  source: '',
}

export interface Position {
  line: number
  column: number
  offset: number
}

export interface SourceLocation {
  start: Position
  end: Position
  source: string
}

export interface Node {
  type: NodeTypes
  loc: SourceLocation
}

export interface RootNode extends Node {
  children: TemplateChildNode[]
  source: string
}

export interface TextNode extends Node {
  type: NodeTypes.TEXT
}

export interface ELementNode extends Node {
  type: NodeTypes.ELEMENT
}

export function createRoot(
  children: TemplateChildNode[],
  source = ''
): RootNode {
  return {
    type: NodeTypes.ROOT,
    loc: locStub,
    children,
    source,
  }
}
