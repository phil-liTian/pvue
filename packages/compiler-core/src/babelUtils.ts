/*
 * @Author: phil
 * @Date: 2025-09-24 14:45:03
 */
import { walk } from 'estree-walker'
import type { Node, Identifier } from '@babel/types'

/**
 * 检查节点是否被引用
 * @param node 当前节点
 * @param parent 父节点
 * @param grandParent 祖父节点
 * @returns 是否被引用
 */
function isReferenced(node: Node, parent: Node, grandParent: Node): boolean {
  switch (parent.type) {
    case 'MemberExpression': {
      return parent.object === node
    }
  }

  return true
}

export function isReferencedIdentifier(
  id: Identifier,
  parent: Node | null,
  parentStack: Node[]
) {
  if (!parent) return true

  if (isReferenced(id, parent, parentStack[parentStack.length - 2])) {
    return true
  }

  return false
}

export function walkIdentifiers(
  root: Node,
  onIdentifier: (
    node: Identifier,
    parent: Node | null,
    parentStack: Node[],
    isReference: boolean
  ) => void,
  includeAll = false,
  parentStack: Node[] = []
) {
  // @ts-ignore
  walk(root, {
    enter(node: Node, parent: Node | null) {
      // console.log('node', node.type)

      if (node.type === 'Identifier') {
        const isRefed = isReferencedIdentifier(node, parent, parentStack)

        onIdentifier(node, parent, parentStack, isRefed)
      }
    },
    leave(node) {
      // console.log('leave node', node.type)
    },
  })
}
