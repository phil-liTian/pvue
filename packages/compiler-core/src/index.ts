export { baseParse } from './parser'

export type { Node, Position, SourceLocation, RootNode, TextNode } from './ast'
export { NodeTypes, locStub } from './ast'

export { createRoot } from './ast'

export {
  CREATE_VNODE,
  RESOLVE_DIRECTIVE,
  helperNameMap,
} from './runtimeHelpers'

export { generate } from './codegen'

export { baseCompile } from './compile'

export { transform } from './transform'

export type { CodegenOptions } from './options'
