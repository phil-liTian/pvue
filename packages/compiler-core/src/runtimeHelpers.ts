/*
 * @Author: phil
 * @Date: 2025-09-14 20:56:01
 */
export const FRAGMENT: unique symbol = Symbol('Fragment')

export const CREATE_VNODE: unique symbol = Symbol('createVNode')
export const RESOLVE_DIRECTIVE: unique symbol = Symbol('resolveDirective')
export const TO_DISPLAY_STRING: unique symbol = Symbol('toDisplayString')
export const CREATE_COMMENT: unique symbol = Symbol('createCommentVNode')
export const RENDER_SLOT: unique symbol = Symbol('createSlots')
export const CREATE_TEXT: unique symbol = Symbol('createTextVNode')
export const CREATE_ELEMENT_VNODE: unique symbol = Symbol('createElementVNode')
export const OPEN_BLOCK: unique symbol = Symbol('openBlock')
export const CREATE_BLOCK: unique symbol = Symbol('createBlock')
export const CREATE_ELEMENT_BLOCK: unique symbol = Symbol('createElementBlock')

export const helperNameMap: Record<symbol, string> = {
  [CREATE_VNODE]: `createVNode`,
  [RESOLVE_DIRECTIVE]: `resolveDirective`,
  [TO_DISPLAY_STRING]: 'toDisplayString',
  [CREATE_TEXT]: 'createTextVNode',
  [FRAGMENT]: 'Fragment',
  [CREATE_ELEMENT_VNODE]: 'createElementVNode',
  [OPEN_BLOCK]: 'openBlock',
  [CREATE_BLOCK]: 'createBlock',
  [CREATE_ELEMENT_BLOCK]: 'createElementBlock',
}
