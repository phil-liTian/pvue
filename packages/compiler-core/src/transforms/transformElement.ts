import { createVNodeCall, NodeTypes, VNodeCall } from '../ast'
import { NodeTransform } from '../transform'

export const transformElement: NodeTransform = (node, context) => {
  return function postTransformElement() {
    node = context.currentNode

    if (!(node.type === NodeTypes.ELEMENT)) return
    const { tag } = node

    let vnodeTag = `"${tag}"`
    let vnodeProps: VNodeCall['props']
    let vnodeChildren: VNodeCall['children']

    // children
    if (node.children.length) {
      vnodeChildren = node.children
    }

    node.codegenNode = createVNodeCall(
      context,
      vnodeTag,
      vnodeProps,
      vnodeChildren
    )
  }
}
