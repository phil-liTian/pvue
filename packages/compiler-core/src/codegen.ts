import { isArray, isString, isSymbol, PatchFlagNames } from '@pvue/shared'
import {
  CallExpression,
  CompoundExpressionNode,
  FunctionExpression,
  getVNodeBlockHelper,
  getVNodeHelper,
  InterpolationNode,
  JSChildNode,
  NodeTypes,
  RootNode,
  SimpleExpressionNode,
  TemplateChildNode,
  TextNode,
  VNodeCall,
} from './ast'
import { CodegenOptions } from './options'
import { helperNameMap, OPEN_BLOCK, TO_DISPLAY_STRING } from './runtimeHelpers'
import { isText } from './utils'

type CodegenNode = JSChildNode | TemplateChildNode

enum NewlineType {
  Start = 0,
  End = -1,
  None = -2,
  Unknown = -3,
}

export interface CodegenResult {
  code: string
}

export interface CodegenContext extends Omit<CodegenOptions, ''> {
  source: string
  code: string
  indentLevel: number
  push(code: string, newlineIndex?: number): void
  newline(): void
  indent(): void
  deindent(withoutNewLine?: boolean): void
  helper(key: symbol): string
}

const aliasHelper = (s: symbol) => `${helperNameMap[s]}: _${helperNameMap[s]}`

function genModulePreamble(ast: RootNode, context: CodegenContext) {
  const { helpers } = ast
  const { push, runtimeModuleName, newline } = context

  if (ast.helpers.size) {
    const helpers = Array.from(ast.helpers)

    push(
      `import { ${helpers
        .map(s => `${helperNameMap[s]} as _${helperNameMap[s]}`)
        .join(', ')} } from ${JSON.stringify(runtimeModuleName)}\n`,
      NewlineType.End
    )
  }

  newline()

  push('export ')
}

function genFunctionPreamble(ast: RootNode, context: CodegenContext) {
  const {
    newline,
    push,
    runtimeModuleName,
    runtimeGlobalName,
    prefixIdentifiers,
  } = context

  const VueBinding = runtimeGlobalName

  const helpers = Array.from(ast.helpers)
  if (helpers.length) {
    if (prefixIdentifiers) {
      push(`const { ${helpers.map(aliasHelper)} } = ${VueBinding}\n`)
    } else {
      push(`const _PVue = ${VueBinding}\n`, NewlineType.End)
    }
  }

  newline()

  push('return ')
}

function createCodegenContext(
  ast: RootNode,
  {
    mode = 'function',
    runtimeModuleName = 'pvue',
    runtimeGlobalName = 'PVue',
    prefixIdentifiers,
  }: CodegenOptions
): CodegenContext {
  const context: CodegenContext = {
    mode,
    source: '',
    code: ``,
    indentLevel: 0,
    runtimeModuleName,
    runtimeGlobalName,
    prefixIdentifiers,
    push(code) {
      context.code += code
    },

    helper(key) {
      return `_${helperNameMap[key]}`
    },

    // 缩进并且换行
    indent() {
      newline(++context.indentLevel)
    },

    // 回撤并且换行
    deindent(withoutNewLine = false) {
      if (withoutNewLine) {
      } else {
        newline(--context.indentLevel)
      }
    },

    // 换行
    newline() {
      newline(context.indentLevel)
    },
  }

  function newline(n: number) {
    context.push('\n' + `  `.repeat(n))
  }

  return context
}

export function generate(
  ast: RootNode,
  options: CodegenOptions = {}
): CodegenResult {
  const context = createCodegenContext(ast, options)

  const { mode, push, indent, deindent, newline, prefixIdentifiers } = context

  const useWithBlock = mode === 'function' && !prefixIdentifiers

  const helpers = Array.from(ast.helpers)
  const hasHelpers = helpers.length > 0

  // mode指定为module
  if (mode === 'module') {
    genModulePreamble(ast, context)
  } else {
    //
    genFunctionPreamble(ast, context)
  }

  const functionName = 'render'
  const args = ['_ctx', '_cache']

  const signature = args.join(', ')

  push(`function ${functionName}(${signature}) {`)
  indent()

  // push('console.log("_ctx",_ctx)\n')

  if (useWithBlock) {
    push('with (_ctx) {')
    indent()

    if (hasHelpers) {
      push(`const { ${helpers.map(aliasHelper).join(', ')} } = _PVue\n`)
      // push('console.log("a",_toDisplayString())\n')

      newline()
    }
  }

  push('return ')

  if (ast.codegenNode) {
    genNode(ast.codegenNode, context)
  }

  // test
  if (useWithBlock) {
    deindent()
    push('}')
  }

  deindent()
  push('}')

  return {
    code: context.code,
  }
}

function genNode(node: CodegenNode, context: CodegenContext) {
  const { push, helper } = context

  if (isSymbol(node)) {
    push(helper(node))
    return
  }

  switch (node.type) {
    // 处理for循环
    // 处理元素节点
    case NodeTypes.ELEMENT:
    case NodeTypes.FOR: {
      genNode(node.codegenNode, context)
      break
    }

    case NodeTypes.TEXT_CALL: {
      genNode(node.codegenNode, context)
      break
    }

    case NodeTypes.JS_CALL_EXPRESSION: {
      genCallExpression(node, context)
      break
    }

    case NodeTypes.SIMPLE_EXPRESSION: {
      genExpression(node, context)
      break
    }

    // 处理插值
    case NodeTypes.INTERPOLATION: {
      genInterpolation(node, context)
      break
    }

    // 处理混合类型表达式
    case NodeTypes.COMPOUND_EXPRESSION: {
      genCompoundExpression(node, context)
      break
    }

    // 处理文本
    case NodeTypes.TEXT: {
      genText(node, context)
      break
    }

    case NodeTypes.VNODE_CALL: {
      // 多children节点, 或者
      genVNodeCall(node, context)
      break
    }

    // 处理函数
    case NodeTypes.JS_FUNCTION_EXPRESSION: {
      genFunctionExpression(node, context)
      break
    }
  }
}

function genExpression(node: SimpleExpressionNode, context: CodegenContext) {
  const { content, isStatic } = node
  context.push(content)
}

function genText(node: TextNode, context: CodegenContext) {
  context.push(JSON.stringify(node.content))
}

// 处理插值
function genInterpolation(node: InterpolationNode, context: CodegenContext) {
  const { push, helper } = context
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(')')
}

// 处理JS_CALL_EXPRESSION类型的node节点 TEXT_CALL的codegenNode就是JS_CALL_EXPRESSION类型的
function genCallExpression(node: CallExpression, context: CodegenContext) {
  const { helper, push } = context
  const { callee } = node
  const _callee = isString(callee) ? callee : helper(callee)
  push(`${_callee}(`)
  genNodeList(node.arguments, context)
  push(')')
}

// 处理混合表达式
function genCompoundExpression(
  node: CompoundExpressionNode,
  context: CodegenContext
) {
  const { push } = context
  for (let i = 0; i < node.children.length; i++) {
    const child = node.children[i]

    if (isString(child)) {
      push(child)
    } else {
      genNode(child, context)
    }
  }
}

// 处理VNODE_CALL 多根节点 或者 type是elment 并且 tagType是element的节点都是VNODE_CALL类型的
function genVNodeCall(node: VNodeCall, context: CodegenContext) {
  const { push, helper } = context
  const {
    isBlock,
    isComponent,
    tag,
    props,
    children,
    dynamicProps,
    disableTracking,
    patchFlag,
  } = node

  if (isBlock) {
    push(`(${helper(OPEN_BLOCK)}(${disableTracking ? 'true' : ''}), `)
  }

  const callHelper: symbol = isBlock
    ? getVNodeBlockHelper(isComponent)
    : getVNodeHelper(isComponent)

  push(`${helper(callHelper)}(`)

  // 处理children节点
  let patchFlagString

  if (patchFlag) {
    if (patchFlag < 0) {
    } else {
      const flagNames = Object.keys(PatchFlagNames)
        .map(Number)
        .filter(n => n > 0 && n & patchFlag)
        .map(n => PatchFlagNames[n])
        .join(', ')
      patchFlagString = patchFlag + ` /* ${flagNames} */`
    }
  }

  genNodeList(
    genNullableArgs([tag, props, children, patchFlagString, dynamicProps]),
    context
  )
  push(`)`)

  if (isBlock) {
    push(')')
  }
}

function genNodeListAsArray(nodes, context: CodegenContext) {
  const { push, indent, deindent } = context
  const multilines = nodes.length > 3 || nodes.some(n => !isText(n))

  push('[')
  multilines && indent()
  genNodeList(nodes, context, multilines)
  multilines && deindent()
  push(']')
}

function genNodeList(
  nodes: (string | TemplateChildNode | Symbol | CodegenNode)[],
  context: CodegenContext,
  multilines: boolean = false,
  comma: boolean = true
) {
  const { push, newline } = context
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i]
    if (isString(node)) {
      push(node)
    } else if (isArray(node)) {
      // node里面的元素有可能还是一个array
      genNodeListAsArray(node, context)
    } else {
      // symbol
      genNode(node, context)
    }

    if (i < nodes.length - 1) {
      if (multilines) {
        comma && push(',')
        newline()
      } else {
        comma && push(', ')
      }
    }
  }
}

// 处理函数
function genFunctionExpression(
  node: FunctionExpression,
  context: CodegenContext
) {
  const { push, indent, deindent } = context
  const { params, newline, returns } = node

  push('(')
  console.log('parase', params)
  if (isArray(params)) {
    genNodeList(params, context)
  }

  push(') => ')

  if (newline) {
    push('{')
    indent()
  }

  if (returns) {
    if (newline) {
      push('return ')
    }
  }

  if (newline) {
    deindent()
    push('}')
  }
}

/**
 * 生成可为空的参数数组
 * @param args 输入参数数组
 * @returns 处理后的参数数组，移除末尾的null/undefined并将剩余null/undefined替换为'null'
 */
function genNullableArgs(args: any[]): CallExpression['arguments'] {
  let i = args.length
  while (i--) {
    if (args[i] != null) break
  }
  return args.slice(0, i + 1).map(arg => arg || 'null')
}
