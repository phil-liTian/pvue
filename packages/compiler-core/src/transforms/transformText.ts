/*
 * @Author: phil
 * @Date: 2025-09-19 11:09:28
 */
import { PatchFlagNames, PatchFlags } from '@pvue/shared'
import {
  CallExpression,
  CompoundExpressionNode,
  ConstantTypes,
  createCallExpression,
  createCompoundExpression,
  NodeTypes,
} from '../ast'
import { CREATE_TEXT } from '../runtimeHelpers'
import { NodeTransform } from '../transform'
import { isText } from '../utils'
import { getConstantType } from './cacheStatic'

export const transformText: NodeTransform = (node, context) => {
  if (node.type === NodeTypes.ROOT || node.type === NodeTypes.FOR) {
    return () => {
      const children = node.children

      let currentContainer: CompoundExpressionNode | undefined = undefined
      for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isText(child)) {
          for (let j = i + 1; j < children.length; j++) {
            const next = children[j]
            if (isText(next)) {
              if (!currentContainer) {
                currentContainer = children[i] = createCompoundExpression(
                  [child],
                  child.loc
                )
              }

              currentContainer.children.push(` + `, next)

              children.splice(j, 1)
              j--
            } else {
              // 如果next不是 text类型的 则需要当作下一个children处理, 进入下一次i循环
              currentContainer = undefined
              break
            }
          }
        }
      }

      // 处理后的children只有一个的话 就不需要将当前节点处理成TEXT_CALL了
      if (children.length === 1 && node.type === NodeTypes.ROOT) {
        return
      }

      // <div/>{{ foo }} bar {{ baz }}<div/>
      // 如果children里面的元素 是text 或者是 COMPOUND_EXPRESSION类型的节点 需要特殊处理
      for (let i = 0; i < children.length; i++) {
        const child = children[i]
        if (isText(child) || child.type === NodeTypes.COMPOUND_EXPRESSION) {
          let callArgs: CallExpression['arguments'] = []

          if (child.type !== NodeTypes.TEXT || child.content !== ' ') {
            callArgs.push(child)
          }

          if (getConstantType(child, context) === ConstantTypes.NOT_CONSTANT) {
            callArgs.push(
              PatchFlags.TEXT +
                (__DEV__ ? ` /* ${PatchFlagNames[PatchFlags.TEXT]} */` : '')
            )
          }

          children[i] = {
            type: NodeTypes.TEXT_CALL,
            codegenNode: createCallExpression(
              context.helper(CREATE_TEXT),
              callArgs
            ),
          }
        }
      }
    }
  }
}
