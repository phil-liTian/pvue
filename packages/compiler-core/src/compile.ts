import { extend, isString } from '@pvue/shared'
import { RootNode } from './ast'
import { baseParse } from './parser'
import { generate } from './codegen'
import { transform } from './transform'
import { transformElement } from './transforms/transformElement'
import { transformText } from './transforms/transformText'

export function getBaseTransformPreset() {
  return [[transformElement, transformText]]
}

export function baseCompile(source: string | RootNode) {
  const ast = isString(source) ? baseParse(source) : source
  const [nodeTransforms] = getBaseTransformPreset()

  transform(
    ast,
    extend(
      {},
      {
        nodeTransforms: [...nodeTransforms],
      }
    )
  )
  return generate(ast, {})
}
