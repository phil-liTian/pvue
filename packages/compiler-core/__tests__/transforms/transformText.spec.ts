import { PatchFlags } from '@pvue/shared'
import {
  CompilerOptions,
  NodeTypes,
  generate,
  baseParse as parse,
  transform,
} from '../../src/index'
import { CREATE_TEXT } from '../../src/runtimeHelpers'
import { transformText } from '../../src/transforms/transformText'
import { transformElement } from '../../src/transforms/transformElement'
import { genFlagText } from '../testUtils'

function transformWithTextOpt(template: string, options: CompilerOptions = {}) {
  const ast = parse(template)
  transform(ast, {
    nodeTransforms: [
      // transformFor,
      // ...(options.prefixIdentifiers ? [transformExpression] : []),
      transformElement,
      transformText,
    ],
    ...options,
  })
  return ast
}

describe('compiler: transform text', () => {
  test('no consecutive text', () => {
    const root = transformWithTextOpt(`{{ foo }}`)
    expect(root.children[0]).toMatchObject({
      type: NodeTypes.INTERPOLATION,
      content: {
        content: `foo`,
      },
    })

    expect(generate(root).code).toMatchSnapshot()
  })

  test('consecutive text', () => {
    const root = transformWithTextOpt(`{{ foo }} bar {{ baz }}`)
    expect(root.children.length).toBe(1)

    expect(root.children[0]).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { type: NodeTypes.INTERPOLATION, content: { content: `foo` } },
        ` + `,
        { type: NodeTypes.TEXT, content: ` bar ` },
        ` + `,
        { type: NodeTypes.INTERPOLATION, content: { content: `baz` } },
      ],
    })

    expect(generate(root).code).toMatchSnapshot()
  })

  test('consecutive text between elements', () => {
    const root = transformWithTextOpt(`<div/>{{ foo }} bar {{ baz }}<div/>`)
    expect(root.children.length).toBe(3)
    expect(root.children[0].type).toBe(NodeTypes.ELEMENT)
    expect(root.children[1]).toMatchObject({
      // when mixed with elements, should convert it into a text node call
      type: NodeTypes.TEXT_CALL,
      codegenNode: {
        type: NodeTypes.JS_CALL_EXPRESSION,
        callee: CREATE_TEXT,
        arguments: [
          {
            type: NodeTypes.COMPOUND_EXPRESSION,
            children: [
              { type: NodeTypes.INTERPOLATION, content: { content: `foo` } },
              ` + `,
              { type: NodeTypes.TEXT, content: ` bar ` },
              ` + `,
              { type: NodeTypes.INTERPOLATION, content: { content: `baz` } },
            ],
          },
          genFlagText(PatchFlags.TEXT),
        ],
      },
    })
    expect(root.children[2].type).toBe(NodeTypes.ELEMENT)
    expect(generate(root).code).toMatchSnapshot()
  })
})
