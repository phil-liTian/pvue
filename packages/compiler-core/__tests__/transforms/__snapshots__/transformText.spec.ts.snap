// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compiler: transform text > <template v-for> 1`] = `
"const _PVue = PVue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { renderList: _renderList, Fragment: _Fragment, openBlock: _openBlock, createElementBlock: _createElementBlock, createTextVNode: _createTextVNode } = _PVue

    return (_openBlock(true), _createElementBlock(_Fragment, null, _renderList(list, (i) => {
      return (_openBlock(), _createElementBlock(_Fragment, null, [
        _createTextVNode("foo")
      ], 64 /* STABLE_FRAGMENT */))
    }), 256 /* UNKEYED_FRAGMENT */))
  }
}"
`;

exports[`compiler: transform text > consecutive text 1`] = `
"const _PVue = PVue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { toDisplayString: _toDisplayString } = _PVue

    return _toDisplayString(foo) + " bar " + _toDisplayString(baz)
  }
}"
`;

exports[`compiler: transform text > consecutive text between elements 1`] = `
"const _PVue = PVue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString, createTextVNode: _createTextVNode, Fragment: _Fragment, openBlock: _openBlock, createElementBlock: _createElementBlock } = _PVue

    return (_openBlock(), _createElementBlock(_Fragment, null, [
      _createElementVNode("div"),
      _createTextVNode(_toDisplayString(foo) + " bar " + _toDisplayString(baz), 1 /* TEXT */),
      _createElementVNode("div")
    ], 64 /* STABLE_FRAGMENT */))
  }
}"
`;

exports[`compiler: transform text > consecutive text mixed with elements 1`] = `
"const _PVue = PVue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString, createTextVNode: _createTextVNode, Fragment: _Fragment, openBlock: _openBlock, createElementBlock: _createElementBlock } = _PVue

    return (_openBlock(), _createElementBlock(_Fragment, null, [
      _createElementVNode("div"),
      _createTextVNode(_toDisplayString(foo) + " bar " + _toDisplayString(baz), 1 /* TEXT */),
      _createElementVNode("div"),
      _createTextVNode("hello"),
      _createElementVNode("div")
    ], 64 /* STABLE_FRAGMENT */))
  }
}"
`;

exports[`compiler: transform text > no consecutive text 1`] = `
"const _PVue = PVue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { toDisplayString: _toDisplayString } = _PVue

    return _toDisplayString(foo)
  }
}"
`;

exports[`compiler: transform text > text between elements (static) 1`] = `
"const _PVue = PVue

return function render(_ctx, _cache) {
  with (_ctx) {
    const { createElementVNode: _createElementVNode, createTextVNode: _createTextVNode, Fragment: _Fragment, openBlock: _openBlock, createElementBlock: _createElementBlock } = _PVue

    return (_openBlock(), _createElementBlock(_Fragment, null, [
      _createElementVNode("div"),
      _createTextVNode("hello"),
      _createElementVNode("div")
    ], 64 /* STABLE_FRAGMENT */))
  }
}"
`;

exports[`compiler: transform text > with prefixIdentifiers: true 1`] = `
"const { toDisplayString: _toDisplayString } = PVue

return function render(_ctx, _cache) {
  return _toDisplayString(_ctx.foo) + " bar " + _toDisplayString(_ctx.baz + _ctx.qux)
}"
`;
